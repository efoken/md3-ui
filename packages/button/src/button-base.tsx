import { useEventCallback, useForkRef } from "@md3-ui/hooks"
import {
  OwnerStateProps,
  styled,
  StyleSheet,
  SxProps,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { splitProps, __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  Platform,
  Pressable as RNPressable,
  PressableProps as RNPressableProps,
  StyleProp,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"

export type ButtonBaseProps = Omit<
  RNPressableProps,
  "children" | "disabled" | "style"
> & {
  /** @default false */
  centerRipple?: boolean
  children?: React.ReactNode
  /** @default false */
  disabled?: boolean
  /** @default false */
  disableRipple?: boolean
  focusColor?: string
  /** @default 0.12 */
  focusOpacity?: string
  hoverColor?: string
  /** @default 0.08 */
  hoverOpacity?: number
  pressedColor?: string
  /** @default 0.12 */
  pressedOpacity?: number
  rippleColor?: string
  style?: StyleProp<RNViewStyle>
  styles?: {
    root?: RNViewStyle
    container?: RNViewStyle
  }
  sx?: SxProps
}

export type ButtonBaseStyleKey = keyof NonNullable<ButtonBaseProps["styles"]>

const ButtonBaseRoot = styled(RNPressable, {
  name: "ButtonBase",
  slot: "Root",
})<OwnerStateProps<Pick<ButtonBaseProps, "disabled">>>(
  ({ ownerState }) =>
    Platform.OS === "web" && {
      cursor: ownerState.disabled ? "default" : "pointer",
      outlineWidth: 0,
      overflow: "hidden",
      transition: "box-shadow 200ms linear",
      userSelect: "none",
    },
)

const ButtonBaseContainer = styled(RNView, {
  name: "ButtonBase",
  slot: "Container",
})({
  backgroundColor: "rgba(255, 255, 255, 0)",
  overflow: "hidden",
})

const ButtonBaseHover = styled("span", {
  name: "ButtonBase",
  slot: "Hover",
})({
  bottom: 0,
  left: 0,
  position: "absolute",
  right: 0,
  top: 0,
  transition: "opacity 200ms linear",
  zIndex: -1,
})

const ButtonBaseRippleContainer = styled("span", {
  name: "ButtonBase",
  slot: "RippleContainer",
})()

const ButtonBaseRipple = styled("span", {
  name: "ButtonBase",
  slot: "Ripple",
})<
  OwnerStateProps<
    Required<
      Pick<ButtonBaseProps, "hoverOpacity" | "pressedColor" | "pressedOpacity">
    >
  >
>(({ ownerState, theme }) => ({
  animationDuration: "550ms",
  animationKeyframes: {
    "0%": {
      opacity: 1,
      transform: [{ scale: 0 }],
    },
    "100%": {
      opacity: 0,
      transform: [{ scale: 1 }],
    },
  },
  animationTimingFunction: "ease-in",
  backgroundColor: theme.utils.rgba(
    ownerState.pressedColor,
    ownerState.pressedOpacity - ownerState.hoverOpacity,
  ),
  borderRadius: 999,
  pointerEvents: "none",
  position: "absolute",
  zIndex: -1,
}))

export const ButtonBase = React.forwardRef<RNView, ButtonBaseProps>(
  (inProps, ref) => {
    const {
      centerRipple = false,
      children,
      disabled = false,
      disableRipple = false,
      focusColor: focusColorProp,
      focusOpacity = 0.12,
      hoverColor: hoverColorProp,
      hoverOpacity = 0.08,
      href,
      onKeyDown,
      onKeyUp,
      onPress,
      pressedColor: pressedColorProp,
      pressedOpacity = 0.12,
      rippleColor,
      style,
      styles,
      ...props
    } = useThemeProps({ name: "ButtonBase", props: inProps })

    const theme = useTheme()

    const rootRef = React.useRef<RNView>(null)
    const handleRef = useForkRef(rootRef, ref)

    const [ripples, setRipples] = React.useState<
      { id: number; ripple: React.ReactElement }[]
    >([])

    const [{ backgroundColor = null, ...containerStyle } = {}] = splitProps(
      StyleSheet.flatten(style),
      [
        "backgroundColor",
        "borderRadius",
        "bottom",
        "height",
        "left",
        "margin",
        "marginBottom",
        "marginEnd",
        "marginHorizontal",
        "marginLeft",
        "marginRight",
        "marginStart",
        "marginTop",
        "marginVertical",
        "position",
        "right",
        "top",
        "width",
      ],
    )

    const hoverColor = rippleColor ?? hoverColorProp ?? theme.color.background
    const focusColor = rippleColor ?? focusColorProp ?? hoverColor
    const pressedColor =
      rippleColor ?? pressedColorProp ?? theme.color.background

    const underlayColor =
      // When button has no background color, simply use `rippleColor` as is
      backgroundColor == null
        ? theme.utils.rgba(pressedColor, pressedOpacity)
        : // Otherwise we mix the `rippleColor` with the background color
          theme.utils.mix(
            pressedOpacity,
            pressedColor,
            backgroundColor.toString(),
          )

    const removeRipple = (rippleID: number) => () => {
      setRipples((prevRipples) => prevRipples.filter((r) => r.id !== rippleID))
    }

    const appendRipple = React.useCallback(
      (event: TouchEvent | MouseEvent | {} = {}) => {
        const currentRef = rootRef.current as HTMLDivElement | null

        if (disabled || disableRipple || currentRef == null) {
          return
        }

        const { width, height, left, top } = currentRef.getBoundingClientRect()

        // Get the size of the ripple
        let rippleX: number
        let rippleY: number
        let rippleSize: number

        if (
          centerRipple ||
          (!("clientX" in event) && !("touches" in event)) ||
          ("clientX" in event && event?.clientX === 0 && event?.clientY === 0)
        ) {
          rippleX = Math.round(width / 2)
          rippleY = Math.round(height / 2)
        } else {
          const { clientX, clientY } =
            "touches" in event ? event.touches[0] : event
          rippleX = Math.round(clientX - left)
          rippleY = Math.round(clientY - top)
        }

        if (centerRipple) {
          rippleSize = Math.sqrt((2 * width ** 2 + height ** 2) / 3)
        } else {
          const sizeX =
            Math.max(Math.abs(currentRef.clientWidth - rippleX), rippleX) * 2 +
            2
          const sizeY =
            Math.max(Math.abs(currentRef.clientHeight - rippleY), rippleY) * 2 +
            2
          rippleSize = Math.sqrt(sizeX ** 2 + sizeY ** 2)
        }

        const rippleID =
          ripples.length === 0 ? 1 : ripples[ripples.length - 1].id + 1

        const ripple = (
          <ButtonBaseRipple
            key={rippleID}
            ownerState={{
              hoverOpacity,
              pressedColor,
              pressedOpacity,
            }}
            style={{
              height: rippleSize,
              left: -(rippleSize / 2) + rippleX,
              top: -(rippleSize / 2) + rippleY,
              width: rippleSize,
            }}
            onAnimationEnd={removeRipple(rippleID)}
          />
        )

        setRipples((prevRipples) => [...prevRipples, { id: rippleID, ripple }])
      },
      [
        rootRef,
        disabled,
        disableRipple,
        centerRipple,
        ripples,
        pressedColor,
        pressedOpacity,
        hoverOpacity,
      ],
    )

    const isNonNativeButton = () => {
      const buttonEl = rootRef.current as HTMLElement | null
      return (
        buttonEl?.tagName !== "BUTTON" &&
        !(buttonEl?.tagName === "A" && (buttonEl as HTMLAnchorElement).href)
      )
    }

    const handleKeyDown = useEventCallback((event: React.KeyboardEvent) => {
      if (event.key === "Enter" || event.key === " ") {
        appendRipple()
      }

      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === " "
      ) {
        event.preventDefault()
      }

      onKeyDown?.(event)

      // Keyboard accessibility for non interactive elements
      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === "Enter" &&
        !disabled
      ) {
        event.preventDefault()
        onPress?.(event as any)
      }
    })

    const handleKeyUp = useEventCallback((event: React.KeyboardEvent) => {
      onKeyUp?.(event)

      // Keyboard accessibility for non interactive elements
      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === " " &&
        !event.defaultPrevented
      ) {
        onPress?.(event as any)
      }
    })

    React.useEffect(() => {
      const currentRef = rootRef.current as HTMLDivElement | null

      if (Platform.OS === "web" && currentRef != null) {
        currentRef.addEventListener("mousedown", appendRipple)
        currentRef.addEventListener("touchstart", appendRipple)

        return () => {
          currentRef.removeEventListener("mousedown", appendRipple)
          currentRef.removeEventListener("touchend", appendRipple)
        }
      }

      return () => {}
    }, [appendRipple, handleKeyDown, rootRef])

    const ownerState = {
      disabled,
    }

    const button = (
      <ButtonBaseRoot
        ref={handleRef}
        accessibilityRole={href ? undefined : "button"}
        accessibilityState={{ disabled: disabled || undefined }}
        android_ripple={
          disableRipple
            ? {}
            : {
                borderless: true,
                color: pressedColor,
              }
        }
        disabled={disabled}
        href={href}
        ownerState={ownerState}
        pointerEvents={disabled ? "none" : undefined}
        {...(Platform.OS === "web" && {
          style: [style, styles?.root],
        })}
        {...(Platform.OS === "android" && {
          // For Android we need a wrapping View to cut off the ripple effect.
          // Because of this View and when the button should have
          // `position: absolute`, we need to apply it to the wrapper View and
          // override it here to have `position: relative` with 0px inset.
          style: [
            style,
            styles?.root,
            {
              bottom: 0,
              left: 0,
              marginBottom: 0,
              marginEnd: 0,
              marginLeft: 0,
              marginRight: 0,
              marginStart: 0,
              marginTop: 0,
              position: "relative",
              right: 0,
              top: 0,
            },
          ],
        })}
        {...(Platform.OS === "ios" && {
          style: ({ pressed }) => [
            style,
            styles?.root,
            pressed && !disableRipple && { backgroundColor: underlayColor },
          ],
        })}
        onKeyDown={handleKeyDown}
        onKeyUp={handleKeyUp}
        onPress={onPress}
        {...props}
      >
        {({ hovered, focused }) => (
          <>
            {React.Children.only(children)}
            {Platform.OS === "web" && (
              <ButtonBaseHover
                style={{
                  backgroundColor: focused ? focusColor : hoverColor,
                  opacity: focused ? focusOpacity : hovered ? hoverOpacity : 0,
                }}
              />
            )}
            {Platform.OS === "web" && !disableRipple && (
              <ButtonBaseRippleContainer>
                {ripples.map(({ ripple }) => ripple)}
              </ButtonBaseRippleContainer>
            )}
          </>
        )}
      </ButtonBaseRoot>
    )

    return Platform.OS === "android" ? (
      <ButtonBaseContainer style={[containerStyle, styles?.container]}>
        {button}
      </ButtonBaseContainer>
    ) : (
      button
    )
  },
)

if (__DEV__) {
  ButtonBase.displayName = "ButtonBase"
}
