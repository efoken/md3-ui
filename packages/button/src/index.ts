export { ButtonBase } from "./button-base"
export type {
  ButtonBaseProps,
  ButtonBaseStateCallbackType,
  ButtonBaseStyleKey,
} from "./button-base"
export { ElevatedButton } from "./elevated-button"
export type {
  ElevatedButtonProps,
  ElevatedButtonStyleKey,
} from "./elevated-button"
export { ExtendedFab } from "./extended-fab"
export type { ExtendedFabProps, ExtendedFabStyleKey } from "./extended-fab"
export { Fab } from "./fab"
export type { FabProps, FabStyleKey } from "./fab"
export { FilledButton } from "./filled-button"
export type { FilledButtonProps, FilledButtonStyleKey } from "./filled-button"
export * from "./filled-icon-button"
export * from "./icon-button"
export * from "./outlined-button"
export * from "./outlined-icon-button"
export * from "./text-button"
export * from "./tonal-button"
export * from "./tonal-icon-button"
export * from "./use-button-base-state"
