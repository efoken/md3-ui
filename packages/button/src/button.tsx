import { useBoolean } from "@md3-ui/hooks"
import { Text, TextProvider } from "@md3-ui/layout"
import {
  OwnerStateProps,
  styled,
  SxProps,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  TextStyle as RNTextStyle,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"
import { ButtonBase, ButtonBaseProps } from "./button-base"

export interface ButtonProps extends ButtonBaseProps {
  /**
   * The content of the component.
   */
  children?: React.ReactNode
  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled?: boolean
  /**
   * Element placed before the children.
   */
  icon?: React.ReactElement
  /**
   * Override or extend the styles applied to the component.
   */
  styles?: {
    root?: RNViewStyle
    content?: RNTextStyle
    icon?: RNViewStyle
    label?: RNTextStyle
  }
  /**
   * The system prop that allows defining system overrides as well as
   * additional styles.
   */
  sx?: SxProps
  /**
   * The variant to use.
   * @default "elevated"
   */
  variant?: "elevated" | "filled" | "tonal" | "outlined" | "text"
}

export type ButtonStyleKey = keyof NonNullable<ButtonProps["styles"]>

const ButtonRoot = styled(ButtonBase, {
  name: "Button",
  slot: "Root",
})<OwnerStateProps<Pick<ButtonProps, "disabled" | "icon" | "variant">>>(
  ({ theme, ownerState }) => [
    {
      alignItems: "center",
      borderRadius: 20,
      flexDirection: "row",
      height: 40,
      paddingHorizontal: theme.spacing(3),

      ...(ownerState.variant === "elevated" && {
        ...theme.elevation.level1,
        backgroundColor: theme.color.surface,

        ...(ownerState.disabled && {
          ...theme.elevation.level0,
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level2,
        },

        ":focus": {
          ...theme.elevation.level1,
        },
      }),

      ...(ownerState.variant === "filled" && {
        ...theme.elevation.level0,
        backgroundColor: theme.color.primary,

        ...(ownerState.disabled && {
          ...theme.elevation.level0,
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level1,
        },

        ":focus": {
          ...theme.elevation.level0,
        },
      }),

      ...(ownerState.variant === "tonal" && {
        ...theme.elevation.level0,
        backgroundColor: theme.color["secondary-container"],

        ...(ownerState.disabled && {
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level1,
        },

        ":focus": {
          ...theme.elevation.level0,
        },
      }),

      ...(ownerState.variant === "outlined" && {
        ...theme.elevation.level0,
        backgroundColor: "rgba(255, 255, 255, 0)",
        borderColor: theme.color.outline,
        borderWidth: 1,
        paddingHorizontal: theme.spacing(3) - 1,

        ...(ownerState.disabled && {
          borderColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":focus": {
          borderColor: theme.color.primary,
        },
      }),

      ...(ownerState.variant === "text" && {
        ...theme.elevation.level0,
        minWidth: 48,
        paddingEnd: ownerState.icon ? theme.spacing(2) : theme.spacing(1.5),
        paddingStart: theme.spacing(1.5),
      }),
    },
  ],
)

const ButtonContent = styled(TextProvider, {
  name: "Button",
  slot: "Content",
})<OwnerStateProps<Pick<ButtonProps, "disabled" | "variant">>>(
  ({ theme, ownerState }) => ({
    ...(ownerState.variant === "elevated" && {
      color: theme.color.primary,
    }),

    ...(ownerState.variant === "filled" && {
      color: theme.color["on-primary"],
    }),

    ...(ownerState.variant === "tonal" && {
      color: theme.color["on-secondary-container"],
    }),

    ...(ownerState.variant === "outlined" && {
      color: theme.color.primary,
    }),

    ...(ownerState.variant === "text" && {
      color: theme.color.primary,
    }),

    ...(ownerState.disabled && {
      color: theme.utils.rgba(theme.color["on-surface"], 0.38),
    }),
  }),
)

const ButtonIcon = styled(RNView, {
  name: "Button",
  slot: "Icon",
})<OwnerStateProps<Pick<ButtonProps, "variant">>>(({ theme, ownerState }) => ({
  marginEnd: theme.spacing(1),
  marginStart: ownerState.variant === "text" ? 0 : -theme.spacing(1),
}))

const ButtonLabel = styled(Text, {
  name: "Button",
  slot: "Label",
})(({ theme }) => ({
  ...theme.typescale["label-large"],
  textAlign: "center",
}))

export const Button = React.forwardRef<RNView, ButtonProps>((inProps, ref) => {
  const {
    children,
    disabled = false,
    icon,
    style,
    styles,
    variant = "elevated",
    ...props
  } = useThemeProps({ name: "Button", props: inProps })

  const theme = useTheme()

  const [hovered, handleHover] = useBoolean()
  const [focused, handleFocus] = useBoolean()

  const ownerState = {
    disabled,
    focused,
    hovered,
    icon,
    variant,
  }

  return (
    <ButtonRoot
      ref={ref}
      disabled={disabled}
      ownerState={ownerState}
      rippleColor={
        variant === "elevated"
          ? theme.color.primary
          : variant === "filled"
          ? theme.color["on-primary"]
          : variant === "tonal"
          ? theme.color["on-secondary-container"]
          : variant === "outlined"
          ? theme.color.primary
          : variant === "text"
          ? theme.color.primary
          : undefined
      }
      style={[styles?.root, style]}
      onBlur={handleFocus.off}
      onFocus={handleFocus.on}
      onHoverIn={handleHover.on}
      onHoverOut={handleHover.off}
      {...props}
    >
      <ButtonContent ownerState={ownerState} style={styles?.content}>
        {icon && (
          <ButtonIcon ownerState={ownerState} style={styles?.icon}>
            {React.cloneElement(icon, {
              height: 18,
              width: 18,
            })}
          </ButtonIcon>
        )}
        <ButtonLabel style={styles?.label}>{children}</ButtonLabel>
      </ButtonContent>
    </ButtonRoot>
  )
})

if (__DEV__) {
  Button.displayName = "Button"
}
