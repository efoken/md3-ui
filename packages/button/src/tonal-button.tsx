import { Text } from "@md3-ui/layout"
import {
  OwnerStateProps,
  styled,
  StylesProp,
  SxProps,
  TextStyleProvider,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  Platform,
  TextStyle as RNTextStyle,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"
import { ButtonBase, ButtonBaseProps } from "./button-base"
import { useButtonBaseState } from "./use-button-base-state"

export interface TonalButtonProps extends ButtonBaseProps {
  /**
   * The content of the component.
   */
  children?: React.ReactNode
  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled?: boolean
  /**
   * Element placed before the children.
   */
  icon?: React.ReactElement
  /**
   * Override or extend the styles applied to the component.
   */
  styles?: StylesProp<{
    root?: RNViewStyle
    content?: RNTextStyle
    icon?: RNViewStyle
    label?: RNTextStyle
  }>
  /**
   * The system prop that allows defining system overrides as well as
   * additional styles.
   */
  sx?: SxProps
}

export type TonalButtonStyleKey = keyof NonNullable<TonalButtonProps["styles"]>

type TonalButtonOwnerState = Pick<TonalButtonProps, "disabled"> & {
  focused: boolean
  hovered: boolean
  pressed: boolean
}

const TonalButtonRoot = styled(ButtonBase, {
  name: "TonalButton",
  slot: "Root",
})<OwnerStateProps<TonalButtonOwnerState>>(({ theme, ownerState }) => [
  {
    ...theme.comp.tonalButton.container.elevation,
    alignItems: "center",
    backgroundColor: theme.comp.tonalButton.container.color,
    borderRadius: theme.comp.tonalButton.container.shape,
    flexDirection: "row",
    height: theme.comp.tonalButton.container.height,
    justifyContent: "center",
    paddingHorizontal: 24,

    ...(Platform.OS !== "web" && {
      shadowColor: theme.comp.tonalButton.container.shadowColor,
    }),

    ...(ownerState.disabled && {
      ...theme.comp.tonalButton.disabled.container.elevation,
      backgroundColor: theme.utils.rgba(
        theme.comp.tonalButton.disabled.container.color,
        theme.comp.tonalButton.disabled.container.opacity,
      ),
    }),

    ...(ownerState.hovered && {
      ...theme.comp.tonalButton.hover.container.elevation,
    }),

    ...(ownerState.focused && {
      ...theme.comp.tonalButton.focus.container.elevation,
    }),

    ...(ownerState.pressed && {
      ...theme.comp.tonalButton.pressed.container.elevation,
    }),
  },
])

const TonalButtonContent = styled(TextStyleProvider, {
  name: "TonalButton",
  slot: "Content",
  skipSx: true,
})<OwnerStateProps<TonalButtonOwnerState>>(({ theme, ownerState }) => ({
  color: theme.comp.tonalButton.labelText.color,

  ...(ownerState.disabled && {
    color: theme.utils.rgba(
      theme.comp.tonalButton.disabled.labelText.color,
      theme.comp.tonalButton.disabled.labelText.opacity,
    ),
  }),

  ...(ownerState.hovered && {
    color: theme.comp.tonalButton.hover.labelText.color,
  }),

  ...(ownerState.focused && {
    color: theme.comp.tonalButton.focus.labelText.color,
  }),

  ...(ownerState.pressed && {
    color: theme.comp.tonalButton.pressed.labelText.color,
  }),
}))

const TonalButtonIcon = styled(RNView, {
  name: "TonalButton",
  slot: "Icon",
  skipSx: true,
})({
  marginEnd: 8,
  marginStart: -8,
})

const TonalButtonLabel = styled(Text, {
  name: "TonalButton",
  slot: "Label",
  skipSx: true,
})(({ theme }) => ({
  ...theme.comp.tonalButton.labelText.textStyle,
  textAlign: "center",
}))

export const TonalButton = React.forwardRef<RNView, TonalButtonProps>(
  (inProps, ref) => {
    const {
      children,
      disabled = false,
      icon,
      style,
      styles,
      ...props
    } = useThemeProps({ name: "TonalButton", props: inProps })

    const theme = useTheme()

    const { focused, hovered, pressed, ...buttonBaseProps } =
      useButtonBaseState(props)

    const ownerState = {
      disabled,
      focused,
      hovered,
      pressed,
    }

    return (
      <TonalButtonRoot
        ref={ref}
        disabled={disabled}
        focusColor={theme.comp.tonalButton.focus.stateLayer.color}
        focusOpacity={theme.comp.tonalButton.focus.stateLayer.opacity}
        hoverColor={theme.comp.tonalButton.hover.stateLayer.color}
        hoverOpacity={theme.comp.tonalButton.hover.stateLayer.opacity}
        ownerState={ownerState}
        pressedColor={theme.comp.tonalButton.pressed.stateLayer.color}
        pressedOpacity={theme.comp.tonalButton.pressed.stateLayer.opacity}
        style={[styles?.root, style]}
        {...buttonBaseProps}
        {...props}
      >
        <TonalButtonContent ownerState={ownerState} style={styles?.content}>
          {icon && (
            <TonalButtonIcon style={styles?.icon}>
              {React.cloneElement(icon, {
                height: theme.comp.tonalButton.withIcon.icon.size,
                width: theme.comp.tonalButton.withIcon.icon.size,
              })}
            </TonalButtonIcon>
          )}
          <TonalButtonLabel style={styles?.label}>{children}</TonalButtonLabel>
        </TonalButtonContent>
      </TonalButtonRoot>
    )
  },
)

if (__DEV__) {
  TonalButton.displayName = "TonalButton"
}
