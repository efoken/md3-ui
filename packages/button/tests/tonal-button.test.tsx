import { describeConformance, render, screen } from "@md3-ui/test-utils"
import { TonalButton } from "../src/tonal-button"

describe("<TonalButton />", () => {
  describeConformance(<TonalButton>Conformance</TonalButton>, () => ({
    inheritComponent: "button",
    refInstanceof: window.HTMLDivElement,
    skip: ["asProp"],
  }))

  it("should render", () => {
    render(<TonalButton>Hello World</TonalButton>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render a button with icon", () => {
    render(<TonalButton icon={<span>icon</span>}>Hello World</TonalButton>)
    const button = screen.getByRole("button")
    const icon = button.querySelector('[data-class*="TonalButton-icon"]')

    expect(button).toMatchSnapshot()
    expect(icon).not.toBeNull()
  })

  it("should have a ripple by default", () => {
    render(<TonalButton>Hello World</TonalButton>)
    const button = screen.getByRole("button")

    expect(
      button.querySelector('[data-class*="ButtonBase-rippleContainer"]'),
    ).not.toBeNull()
  })

  it("should disable the ripple", () => {
    render(<TonalButton disableRipple>Hello World</TonalButton>)
    const button = screen.getByRole("button")

    expect(
      button.querySelector('[data-class*="ButtonBase-rippleContainer"]'),
    ).toBeNull()
  })

  it("should automatically change the button to an anchor element when href is provided", () => {
    const { container } = render(
      <TonalButton href="https://google.com">Hello</TonalButton>,
    )
    const button = container.firstChild

    expect(button).toHaveProperty("nodeName", "A")
    expect(button).not.toHaveAttribute("role")
    expect(button).not.toHaveAttribute("type")
    expect(button).toHaveAttribute("href", "https://google.com")
  })
})
