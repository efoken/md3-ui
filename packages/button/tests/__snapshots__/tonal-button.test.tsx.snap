// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TonalButton /> should render 1`] = `
<button
  class="css-view-175oi2r r-touchAction-1otgn73 r-cursor-1loqt21 r-outlineWidth-10paoce r-transition-71lksq r-userSelect-lrvibr r-alignItems-1awozwy r-backgroundColor-oc2s68 r-borderRadius-cpet4d r-boxShadow-cnhr7f r-flexDirection-18u37iz r-height-eu3ka r-justifyContent-1777fci r-paddingInline-cxgwc0"
  data-class="TonalButton-root ButtonBase-root"
  role="button"
  type="button"
>
  <div
    style="color: rgb(29, 25, 43); display: contents;"
  >
    <div
      class="css-text-146c3p1 r-fontFamily-rxj5oj r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="TonalButton-label Text-root"
      dir="auto"
      style="color: rgb(29, 25, 43);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-sp8bj3 r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="background-color: rgb(29, 25, 43); opacity: 0;"
  />
  <span
    class="r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-overflow-1udh08x r-position-u8s1d r-right-zchlnj r-top-ipm5af r-zIndex-417010"
    data-class="ButtonBase-rippleContainer"
  />
</button>
`;

exports[`<TonalButton /> should render a button with icon 1`] = `
<button
  class="css-view-175oi2r r-touchAction-1otgn73 r-cursor-1loqt21 r-outlineWidth-10paoce r-transition-71lksq r-userSelect-lrvibr r-alignItems-1awozwy r-backgroundColor-oc2s68 r-borderRadius-cpet4d r-boxShadow-cnhr7f r-flexDirection-18u37iz r-height-eu3ka r-justifyContent-1777fci r-paddingInline-cxgwc0"
  data-class="TonalButton-root ButtonBase-root"
  role="button"
  type="button"
>
  <div
    style="color: rgb(29, 25, 43); display: contents;"
  >
    <div
      class="css-view-175oi2r r-marginInlineEnd-1wron08 r-marginInlineStart-m5k245"
      data-class="TonalButton-icon"
    >
      <span
        height="18"
        width="18"
      >
        icon
      </span>
    </div>
    <div
      class="css-text-146c3p1 r-fontFamily-rxj5oj r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="TonalButton-label Text-root"
      dir="auto"
      style="color: rgb(29, 25, 43);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-sp8bj3 r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="background-color: rgb(29, 25, 43); opacity: 0;"
  />
  <span
    class="r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-overflow-1udh08x r-position-u8s1d r-right-zchlnj r-top-ipm5af r-zIndex-417010"
    data-class="ButtonBase-rippleContainer"
  />
</button>
`;
